[Lai Han Wen] iP

hanwenlai opened the pull request:
## Duke
> “Your mind is for having ideas, not holding them.” – David Allen

Duke frees your mind of the burden of having to remember things you need to do. It's,
- user-friendly
- text-based
- ~~FAST~~ *LIGHTNING* FAST to use

All you have to do is:
1. download it from this [link](https://github.com/nus-cs2103-AY2223S1/ip).
2. double-click it.
3. add your tasks.
4. let Duke manage your tasks for you 😘 

And most importantly, it is completely **FREE!!!**

Features:
- [x] Adding tasks
- [x] Deleting tasks
- [x] Marking tasks as done
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```
Why wait any longer?? Download it **_now!!!_** 🤪

Commit:
Add Gradle support

hanwenlai committed:
Level-1

hanwenlai committed:
Level-2

hanwenlai committed:
Level-3

hanwenlai committed:
Level-4

hanwenlai committed:
A-TextUiTesting

hanwenlai committed:
Level-5

hanwenlai committed:
Level-6

hanwenlai committed:
Level-6

hanwenlai committed:
branch-Level-7

hanwenlai committed:
branch-Level-8

hanwenlai committed:
Merge branch 'branch-Level-7'

hanwenlai committed:
Merge branch 'branch-Level-8'

hanwenlai committed:
A-MoreOOP

hanwenlai committed:
A-Packages

hanwenlai committed:
Merge remote-tracking branch 'origin/add-gradle-support'

hanwenlai committed:
Add JUnit tests

hanwenlai committed:
Add JavaDoc comments

hanwenlai committed:
Follow coding standard

hanwenlai committed:
Add keyword searching function

hanwenlai committed:
Merge branch 'branch-A-JavaDoc'

hanwenlai committed:
Comply with coding standard

hanwenlai committed:
Add keyword searching function

fungusta commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,25 @@
+ */
+public class DukeException extends Exception {
+
+    public DukeException() {}
```
Noticed a few of these empty constructors, since it does not really do anything, you might want to just leave the class without a constructor

fungusta commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,141 @@
+            if (Objects.equals(action, "T")) {
+                Task t = this.executeTodo(arr[1]);
+                this.ui.showTaskAdded(t, this.tasks);
+                continue;
```
Might want to consider using a switch statement for this to make the code shorter

fungusta commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,140 @@
+                }
+
+                description = s.substring(7, s.indexOf('(') - 1);
+                String time = s.substring(s.indexOf('(') + 5, s.indexOf(')'));
```
Noticed quite a few of these complicated expressions, maybe you can consider following the example in the code quality notes where you do these in steps as seen below
```
int startOfTime = s.indexOf('(') + 5;
int endOfTime = s.indexOf(')');
String time = s.substring(startOfTime, endOfTime);
```

fungusta reviewed (commented):
Nice!! 😄

hanwenlai committed:
Add a GUI to Duke using JavaFX technology.

Dilysss commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    /**
+     * Returns the String representation of a deadline task, where the
```
First sentence of javadocs could be shorter

Dilysss commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,32 @@
+package duke;
+
+import java.io.IOException;
+
```
This line can be deleted!

Dilysss commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,54 @@
+        return this.list.size();
+    }
+
+    public ArrayList<Task> getList() {
+        return this.list;
+    }
+
+    public Task getTask(int i) {
+        return this.list.get(i - 1);
+    }
```
Missing javadocs

Dilysss reviewed (commented):
Overall your code follows the coding standard very well except maybe some missing javadocs. Good job in naming the various methods and variables!

hanwenlai committed:
Refactor various parts of code to improve code quality

Increase readability and clarity of code.

Let's
* replace several nested if-else statements with switch statements
* use named constants in place of "magic numbers"
* extract methods from long methods
* minimize code duplication

hanwenlai committed:
Add Java assertions to various points in the code

Document important assumptions at various points in the code.

Easier to verify correctness of the code.

Let's
* place an assertion at a location that is assumed will not be reached
* use an assertion to assert an internal invariant

hanwenlai committed:
Merge pull request #2 from hanwenlai/branch-A-Assertions

Add Java assertions to various points in the code

hanwenlai committed:
Merge branch 'master' into branch-A-CodeQuality

hanwenlai committed:
Merge pull request #1 from hanwenlai/branch-A-CodeQuality

Refactor various parts of code to improve code quality

hanwenlai committed:
A-CI

Create gradle.yml

hanwenlai committed:
Add the ability to detect duplicate tasks

Prevents user from adding the same task multiple times.

Let's,
* add new methods to compare given task with tasks in saved data
* check if the given task is a duplicate before adding it into the list

hanwenlai committed:
Merge branch 'branch-C-DetectDuplicates'

hanwenlai committed:
Handle more errors that Duke might encounter during usage

Possible errors for user commands "delete", "mark" and "list" were not
handled.

Let's,
* display an error message when user calls "delete", "mark" or "list"
  on an empty task list
* display an error message when user tries to delete or mark a task
  that is out of current task list's range
* display an error message when user calls "delete" or "mark" without
  providing a number

hanwenlai committed:
Add a User Guide

hanwenlai committed:
Add a representative screenshot of the product

hanwenlai committed:
Update inline link in User Guide

hanwenlai committed:
Update User Guide
